### PR TITLE
DRAFT: for debugging a dune issue (sharing Win32-specific C code between coqide and coqidetop)

### PR DESCRIPTION
**Pull request by jfehrle**
NOT FOR REVIEW YET

Here's what I get:

```
$ time (dune build -p coq-core,coqide-server,coqide && dune install coq-core coqide-server coqide)
File "ide/coqide/dune", line 11, characters 0-258:
11 | (executable
12 |  (name idetop)
13 |  (public_name coqidetop.opt)
14 |  (package coqide-server)
15 |  (modules idetop)
16 |   (foreign_stubs
17 |    (language c)
18 |    (names coqide_os_stubs))
19 |  (libraries coq-core.toplevel coqide-server.protocol)
20 |  (modes native byte)
21 |  (link_flags -linkall))
Error: Pure bytecode executables cannot contain foreign stubs.
Hint: If you only need to build a native executable use "(modes exe)".
File "ide/coqide/dune", line 35, characters 9-24:
35 |   (names coqide_os_stubs))
              ^^^^^^^^^^^^^^^
Error: Multiple definitions for the same object file "coqide_os_stubs.o". See
another definition at ide/coqide/dune:18.
Hint: You can avoid the name clash by renaming one of the objects, or by
placing it into a different directory.

real    0m6.714s
user    0m0.000s
sys     0m0.046s
```

I'd like to reuse `coqide_os_specific.ml` and `coqide_os_stubs.c` (both generated, not in git).  I'll need to add a call to `Coqide_os_specific.init` somewhere for the server (as was done in `coqide_main.ml`).

Attn: @rgrinberg

### CODE REVIEW

**Comment by jfehrle** [comment]
PS A more elementary problem: I wanted to share `Coq.get_interrupt_fname` between CoqIDE and the server.  But not understanding the concepts, I don't know what to change in the dune file.  I can create a second copy of the routine as a workaround though that's not good practice.

If I exclude the `dune` file change, I get:

```
File "ide/coqide/idetop.ml", line 33, characters 16-39:
33 |     let fname = Coq.get_interrupt_fname (Unix.getpid ()) in
                     ^^^^^^^^^^^^^^^^^^^^^^^
Error: Unbound module Coq
```

**Comment by jfehrle** [comment]
PPS I tried to sign up for the opam infrastructure mailing list [here](https://lists.ocaml.org/listinfo/infrastructure)  2 days ago, but so far I've only gotten 2 "Delivery incomplete" messages.  If this is now defunct then it should be removed from several web pages.  I forget exactly which pages listed it, but Google is your friend.

**Comment by rgrinberg** [comment]
> Error: Multiple definitions for the same object file "coqide_os_stubs.o". See
another definition at ide/coqide/dune:18.
Hint: You can avoid the name clash by renaming one of the objects, or by
placing it into a different directory.

This error is straightforward I believe. You're trying to use the same stub in two different stanzas, but that's not allowed.

Your options are:

1) Make the stubs live in the library stanza and use the functions in the executables by exposing them in the library's public interface. (I recommend this option)

2) Rename the stubs for the executable stanza. E.g.:

```
(copy coqide_os_stubs coqdie_os_stubs_for_exe)
(executable
   ...
  (foreign_stubs
  (language c)
  (names coqide_os_stubs_for_exe))
```

**Comment by rgrinberg** [comment]
Here's a fix for both issues:
```
diff --git a/ide/coqide/dune b/ide/coqide/dune
index f39e2f5dc7..53f591fee6 100644
--- a/ide/coqide/dune
+++ b/ide/coqide/dune
@@ -13,10 +13,7 @@
  (public_name coqidetop.opt)
  (package coqide-server)
  (modules idetop)
- (foreign_stubs
-  (language c)
-  (names coqide_os_stubs))
- (libraries coq-core.toplevel coqide-server.protocol)
+ (libraries coqide_gui coq-core.toplevel coqide-server.protocol)
  (modes native byte)
  (link_flags -linkall))
```

**Comment by rgrinberg** [comment]
>  A more elementary problem: I wanted to share Coq.get_interrupt_fname between CoqIDE and the server. But not understanding the concepts, I don't know what to change in the dune file. I can create a second copy of the routine as a workaround though that's not good practice.

To fix a problem like this, you start by asking yourself "to which stanza the Coq module belongs to?". By reading the dune file, we determine that it belongs to the library stanza `coqide_gui`. Therefore, we must add `coqide_gui` as a dependency in our executable see this module.

**Comment by jfehrle** [comment]
That works, thanks.

> To fix a problem like this, you start by asking yourself "to which stanza the Coq module belongs to?". By reading the dune file, we determine that it belongs to the library stanza `coqide_gui`.

The underlying concepts of linking and compiling are very widely understood.  How to read a dune file is not.
Dune is not intuitively obvious to a newcomer.  The documentation doesn't explain concepts well. It's hard to find things.  It needs indexing and more hyperlinks.  And it's somewhat short of reference manual style (ie you don't have to read the entire document to find what you need).  Linking in module is a very elementary, common task that should be described.

FWIW, where is "stanza" clearly defined in the doc?  Shouldn't the precise question be "to which library the Coq module belongs to"?  (Guessing that every module must belong to one and only one library.)

File: ide/coqide/dune [hunk 1]
```diff
@@ -13,7 +13,7 @@
  (public_name coqidetop.opt)
  (package coqide-server)
  (modules idetop)
- (libraries coq-core.toplevel coqide-server.protocol)
+ (libraries coqide_gui coq-core.toplevel coqide-server.protocol)
  (modes native byte)
  (link_flags -linkall))
```
Comment (rgrinberg) [inline ide/coqide/dune:18]: I'm curious, why do you need `-linkall` here? Why not let OCaml link only what's needed? You're probably slowing down your build by a good bit doing this.

**Comment by rgrinberg** [comment]
> The underlying concepts of linking and compiling are very widely understood.

Yes, but you should understand the way these are expressed in dune. For example, things existing in the same directory has no bearing in dune like it would with include flags. Instead, dune relies on explicitly specifying dependencies in the `libraries` field.

> FWIW, where is "stanza" clearly defined in the doc? 

A stanza is a toplevel declaration. E.g. `(library ..)`, `(executable ..)`, etc. 

> Shouldn't the precise question be "to which library the Coq module belongs to"? 

Almost. Modules may also belong to executables and may be shared among multiple executables if they are all defined using the same `executables` stanza.

> (Guessing that every module must belong to one and only one library.)

one and only one library or executable(s) stanzas.

**Comment by jfehrle** [comment]
Actually the fix was more complicated than you described.  Including `coqide_gui` apparently causes all files within the library get loaded, which meant that the server began to run the GUI event loop.  I also had to create a new OS specific library because the original OS specific library depended on GUI routines.  Having to explicitly exclude modules (e.g. `(modules (:standard \ ... shared shared_os_specific))` seems awkward.

**Comment by rgrinberg** [comment]
> Having to explicitly exclude modules (e.g. (modules (:standard \ ... shared shared_os_specific)) seems awkward.

To avoid doing that, consider the following hygiene rule:

Only keep one `library`, `executable`, `executables` stanza per directory. Then dune will infer the relationship between modules and stanzas automatically.

File: ide/coqide/dune (hunk 1 above)
Comment (jfehrle) [reply ide/coqide/dune:18]: I have no idea.  Perhaps the author of that line, @ejgallego can explain.

If I didn't have that option then `coqide` wouldn't have linked and loaded the GUI event loop?

**Comment by jfehrle** [comment]
> Yes, but you should understand the way these are expressed in dune.

Eventually, yes.  Not so easy to get there.

File: ide/coqide/dune (hunk 1 above)
Comment (rgrinberg) [reply ide/coqide/dune:18]: You could always pull in the modules that initialize the gui event loop in the executable's main manually. E.g:

```
let () =
  let module M = Gui in ()
```

Comment (jfehrle) [reply ide/coqide/dune:18]: I'm not familiar with `let module`.  Does the OCaml doc describe the construct and explain how it works?

(I would love to see the OCaml documentation rewritten.  I think the doc is a barrier to use of OCaml.)

**Pull request closed without merging**